Feature: Support for React JS server side rendering

mikehearn opened the pull request:
This work in progress PR adds support for rendering ReactJS views server side. It sports:

* The GraalJS engine run in multithreaded mode for high performance. Threads share the JIT compiled code and can render in parallel.
* Javascript can be sandboxed (with the next major version of GraalJS).
* Transparent support for both React and Preact (a lighter weight version of the React runtime).
* Customize the Javascript used to invoke SSR to add features like head managers, or use the prepackaged default scripts to get going straight away.
* Full documentation with tutorial showing you how to prepare your Javascript.
* You can pass any `@Introspectable` Java objects to use as _props_ for your page components. This is convenient for passing in things like the user profile info.
* ~You can integrate fetcher libraries like [SWR](https://swr.vercel.app/) so Micronaut Views React can pre-fetch the API response server-side and then pass it to the client in the initial HTML. The docs show how to integrate this with the `useSWR` hook, which is the recommended way to use this feature. This reduces latency and improves robustness.~
* Logging from Javascript is sent to the Micronaut logs. `console.log` and related will go to the `INFO` level of the logger named `js`, `console.error` and Javascript exceptions will go to the `ERROR` level of the same.

To do list:

* [x] Sandboxing (done but requires an unreleased GraalJS, is off by default as a consequence).
* [x] Upgrade to Micronaut 4.4.0 and get rid of the thread hacks.
* [x] Write more tests to show the sandbox actually works.
* [x] Load the user's JS bundle using the `ResourceLoader` API.
* [ ] Try it out with a more complex React based app, instead of just hello world samples.
* [ ] Work out if there's an easy way to attach a debugger to the server-side JS.
* [x] Document that Micronaut Security will add props to your root object automatically.
* [ ] Work with @sgammon to integrate his excellent work on JS libraries from Elide. This should fix the HTTP fetching implementation.

mikehearn committed:
React Renderer: Add support for server-side rendering using React.

mikehearn committed:
Simplify the code and document it.

mikehearn committed:
Update the Reactonaut docs.

mikehearn committed:
React Renderer: remove the todo list and do a small refactor to prepare for preact support.

mikehearn committed:
React Renderer: add a basic unit test setup

mikehearn committed:
React Renderer: make the user pick the GraalJS edition.

This also removes the need for the weird POM hacks.

mikehearn committed:
React Renderer: fix client.js for the docs

mikehearn committed:
React Renderer: sandboxing isn't yet possible

mikehearn committed:
React Renderer: make path to JS bundles configurable

mikehearn committed:
React Renderer: support Preact

mikehearn committed:
React Renderer: refactor engine logging out.

mikehearn committed:
React Renderer: refactor bundle path computations out.

mikehearn committed:
React Renderer: refactor more of the js out.

mikehearn committed:
React Renderer: more refactorings.

mikehearn committed:
React Renderer: parallelism

mikehearn committed:
React Renderer: update todo list

mikehearn committed:
React Renderer: optimize the reload check.

mikehearn committed:
React Renderer: tweak the name of the global symbol.

mikehearn committed:
React Renderer: make render scripts customizable.

mikehearn committed:
React Renderer: show how to use head managers.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-views?pullRequest=770) <br/>All committers have signed the CLA.

mikehearn committed:
React Renderer: remove docs disclaimer

mikehearn committed:
React Renderer: update client.preact.js to new global symbol name.

graemerocher commented on src/main/docs/guide/views/templates/react.adoc:
```diff
@@ -0,0 +1,28 @@
+TIP: The repository containing this prototype is not yet available. Please contact Mike Hearn in Oracle Labs if you want to try it out.
```
we will need to fix this before merging :)

graemerocher commented on src/main/docs/guide/views/templates/react/reactsettingproperties.adoc:
```diff
@@ -0,0 +1,33 @@
+[configuration]
+----
+micronaut:
+  # HACK
```
remove hack once upgrading to Micronaut 4.4

graemerocher commented on src/main/docs/guide/views/templates/react/reactsettingproperties.adoc:
```diff
@@ -0,0 +1,33 @@
+          cache-seconds: 0
+----
+
+Set the hack property.
```
remove this once on 4.4

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/JSBundlePaths.java:
```diff
@@ -0,0 +1,65 @@
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
try cleanup this code a bit, likely it will fail check style

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/JSContext.java:
```diff
@@ -0,0 +1,146 @@
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+class JSContext implements AutoCloseable {
```
provide a scope for this context. If a new one is needed each time add `Prototype` otherwise `Singleton`

mikehearn replied:
Contexts are pooled, so I'm not sure their lifetime maps to anything the DI container can handle.

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/JSContext.java:
```diff
@@ -0,0 +1,146 @@
+    CompiledJS compiledJS;
+
+    @Inject
+    ReactViewsRendererConfiguration configuration;
```
in generate constructor injection is preferred if possible

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/JSContextPool.java:
```diff
@@ -0,0 +1,45 @@
+ */
+@Singleton
+class JSContextPool {
+    private final LinkedList<SoftReference<JSContext>> contexts = new LinkedList<>();
```
you may want to consider a custom scope for `JSContext`

Also does a concurrent collection make more sense here? or is synchronised required? Perhaps a RentrantReadWriteLock would be more virtual thread friendly

mikehearn replied:
I narrowed the lock slightly but this is the best locking strategy:

- For Loom it's OK because we're not doing any blocking suspends whilst holding the lock.
- A concurrent collection could be used, but has more memory overhead and the lock should not be held for long given that it will normally only loop once or not at all. 

I've profiled this code already and it didn't show up. The overhead of executing the JS drowns everything else.

mikehearn replied:
W.R.T the custom scope, I'm not sure how to implement pooling that way or what benefit it would give. Creating these via the bean context seems easy enough. Is there IoC container support for soft object pooling I could use instead?

graemerocher replied:
ok it is probably fine as is

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/JSEngineLogHandler.java:
```diff
@@ -0,0 +1,53 @@
+        String message = record.getMessage();
+        Throwable thrown = record.getThrown();
+        String level = record.getLevel().getName();
+        switch (level) {
```
use a switch expression

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/OutputStreamToSLF4J.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * An output stream that looks for line separators and then writes out the lines of text to the given logger.
+ */
+class OutputStreamToSLF4J extends OutputStream {
```
can this be final?

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactViewsRenderer.java:
```diff
@@ -0,0 +1,216 @@
+    JSContextPool contextPool;
+
+    @Inject
+    JSBundlePaths jsBundlePaths;
```
use constructor injection

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactViewsRenderer.java:
```diff
@@ -0,0 +1,216 @@
+                var req = HttpRequest.GET(builder.build())
+                    .header(CONTENT_ENCODING, "UTF-8")
+                    .header(ACCEPT, "application/vnd.github.v3+json, application/json")
+                    .header(USER_AGENT, "Micronaut React SSR Prerender");
```
A little bit concerned by the security implications here, where is `urlsToPrefetch` defined? Are they external resources? Seems like we are pre-rendering JS from external sources, could that be a danger? What is the sandboxing model?

If the resource is local can it be retrieved without the the HTTP request the localhost?

mikehearn replied:
The JS doesn't come from external sources. It's all a part of the app, so is "trusted". 

Nonetheless I am working on enabling sandboxing, because of course JavaScript world has giant dependency trees where not everyone maybe is entirely trustworthy, so to protect against supply chain attacks some sandboxing is useful anyway.

> If the resource is local can it be retrieved without the HTTP request the localhost?

See my question in #help on Micronaut Discord. I'm sure it can be, and that would be much better, I just don't know how to do it. I'd need to turn a URL into the right bean and method, and then build a reflective invocation in the same way Micronaut does with a fake HTTP request and all the injected stuff. If you know how to do that, some example code would be great.

graemerocher replied:
where are these resources kept? If they are on the classpath or file system you can use the ResourceResolver API

mikehearn replied:
Well they are usually API endpoints. The idea here is that the template uses the same code to download data on the client as on the server. For example, consider a page that lets you browse the profiles of a list of users. The initial render should send everything including a pre-selected/pre-loaded first user, and then when the user clicks the list, the data of that next user should be loaded. In effect we are simulating what the browser would do, but server side, so optimizing out a round-trip. Thus we can't load the result from the disk. It has to be genuinely simulated, including details that aren't done yet like propagating authentication cookies or headers.

Basically I need a way to efficiently do an HTTP request "as if" it was from the same user that generated the existing one.

graemerocher replied:
hmm, not sure if we have an API for that other than what you are doing, the question is should be allowing URLs other than localhost? Seems dangerous to potentially make outgoing calls and what if the deployed environment doesn't allow them?

mikehearn replied:
Good question.

Frontend JS can and often does make calls to multiple microservices. You can reverse proxy them all of course, and the integrations are hopefully robust to calls failing so this isn't a critical need, but in principle yes you would need to contact non-localhost servers at least sometimes. It probably makes sense to have a configurable whitelist of URL patterns to allow.

sgammon replied:
I think it would be possible to fetch through Micronaut, to allow policies to be applied. Certainly the fetch implementation we use can be customized by policy or redirected to other endpoints.

Food for thought, because JS apps can and do structure their fetching this way. Some serverless database products (i.e. Cloudflare D1) work via REST JSON APIs under the hood, so this is functionally the same for a React app as say, JDBC, if you squint

graemerocher replied:
so if we continue with this approach I would like to see some whitelist implementation. I think from a security perspective it is too risky to by default allow outgoing calls to servers defined by the client JS. This could lead to an avenue whereby some vulnerability on the JS side could penetrate the server.

mikehearn replied:
Of course, it'd be pretty nice to short-circuit the HTTP stack and just pass the returned bean directly into the JS context when that's possible. But I guess that can be an optimization punted to later. You'd need APIs to look up the method for a URL, figure out reflectively if it fits the constraints etc. And presumably do that at compile time.

sgammon replied:
@mikehearn Wait, is this for `fetch` as in fetching external data? Or passing beans? I'll ping you to discuss

mikehearn committed:
React Renderer: add support for sandboxing, currently disabled by default.

GraalJS needs a small fix before this works, so the current version can't be used with the sandbox. The next version should be compatible.

mikehearn committed:
React Renderer: add license headers

mikehearn committed:
React Renderer: fix checkstyle warnings.

mikehearn committed:
React Renderer: resolve a few review comments.

For the rest, I need to do more work on server-side fetching and similar.

mikehearn committed:
React Renderer: add a discussion of why object pooling.

mikehearn commented:
I've been talking with @sgammon about maybe using the code he's got in Elide for this PR. It would mean adding some new MIT licensed dependencies to the React Renderer module. In turn his code depends on the Kotlin standard library and GraalJS. 

The benefits would be that his team has implemented some of the core Web/Node APIs on top of Micronaut already, for example the fetch API. The current PR provides a very MVP/prototype implementation of server side fetching. It's super basic, not parallelized at all, basically expects you to do overrides and hot patches to your code in order to use it. Implementing the fetch API and others would improve compatibility with existing JS, especially for larger projects. Doing this well though is a lot of work, so it makes sense to reuse Sam's effort.

I'll park this work for a week or two, to let Sam work out if/when he'd be able to do this effort. If we can just add a few deps and use them, I'd rather add it all into this one PR. If Sam won't have time, then I might take server-side fetch out of this PR and then look at adding it back in later after extracting the code myself, but that would depend on user demand.

sgammon commented:
I should be able to review and get back to this within a day or two

mikehearn committed:
React Renderer: remove the half-baked fetch support.

It may yet return at some unspecified future date!

mikehearn committed:
React Renderer: update limitations docs.

mikehearn commented:
I've removed the half-baked server side fetch support for now, as that's the part that would change the most given Sam's work. The remaining tasks are small and can hopefully be finished soon. Then it should be ready to merge.

mikehearn committed:
React Renderer: add another unit test for the sandbox.

graemerocher commented on views-react/build.gradle:
```diff
@@ -0,0 +1,86 @@
+}
+
+repositories {
+    mavenLocal()
```
remove mavenLocal() before merge

graemerocher commented on views-react/build.gradle:
```diff
@@ -0,0 +1,86 @@
+    annotationProcessor(mnValidation.micronaut.validation.processor)
+
+    api projects.micronautViewsCore
+    api libs.jetbrains.annotations
```
don't depend on jet brains annotations, we have our own `Nullable` `NonNull` if you need those

graemerocher commented on views-react/build.gradle:
```diff
@@ -0,0 +1,86 @@
+
+    compileOnly(mn.micronaut.management)
+    compileOnly(mnValidation.micronaut.validation)
+    compileOnly(mn.micronaut.http)
```
why are these compileOnly?

mikehearn replied:
I fixed two, left validation because it's only used for `@NotBlank`.

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/CompiledJS.java:
```diff
@@ -0,0 +1,55 @@
+ * into the JVM code cache.
+ */
+@Singleton
+class CompiledJS implements AutoCloseable {
```
annotate with `@Internal`

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/JSBundlePaths.java:
```diff
@@ -0,0 +1,81 @@
+            }
+            return false;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
throw a more concrete exception

mikehearn replied:
What's conventional to use here in Micronaut for IOExceptions from disk?

graemerocher commented on views-react/src/main/java/io/micronaut/views/react/ReactViewsRendererConfiguration.java:
```diff
@@ -0,0 +1,81 @@
+     * chain attacks that might inject code into your server via hijacked React components.
+     * It requires a sufficiently new version of GraalJS. Defaults to OFF.
+     */
+    @Bindable(defaultValue = "false")
```
why is this default `false`? Should be secure by default.

mikehearn replied:
> It requires a sufficiently new version of GraalJS. Defaults to OFF.

Currently released GraalJS does not work with this feature. The 24.1 release should work.

mikehearn replied:
Once the new GraalJS is out, we can follow up with a new PR to change the default setting.

graemerocher commented on src/main/docs/guide/views/templates/react.adoc:
```diff
@@ -0,0 +1,24 @@
+This improves performance by ensuring the page appears before any Javascript has loaded (albeit in a non-responsive
+state) and makes it easier for search engines to index your pages.
+
+Micronaut's support for React SSR has the following useful features:
```
add a `{note}` that this module is experimental and subject to change

mikehearn replied:
I couldn't find an example of the syntax you mean, so wrote a `NOTE: This module is experimental` disclaimer.

mikehearn committed:
React Renderer: more sandbox tests, a bit of refactoring, and expose the page URL to the render script.

mikehearn committed:
React Renderer: address some review comments.

mikehearn commented:
@graemerocher I've done a few improvements in the last couple of commits, including to the docs and I also now expose the request URL to the Javascript (useful for page router components).

W.R.T. using `ResourceLoader`, what's the right way to do it with this API for:

1. Reloading resources automatically if they change on disk, but without polling on every request.
2. Taking into account `viewsConfiguration.getFolder()` as the base?

graemerocher commented:
You can probably use a scheduled job to check for changes and reload I imagine

graemerocher commented:
Actually I think we have a file watch API already. You need to configure the watch paths (see FileWatchConfiguration), then you can write a `@EventListener` for `FileChangedEvent`:

```java
@EventListener
void filesChanged(FileChangedEvent event) {

}
```

mikehearn committed:
React Renderer: use the file watching support.

In practice it's not that important in typical setups, as the build system plugins will restart the whole server anyway. But if you disable that to make things faster, this can help.

mikehearn commented:
File watching support using the standard API is added, though using it requires disabling the build plugin auto-restart behavior of course, as well as configuring it via extra application.properties which the old solution didn't need.

graemerocher commented:
you could disable (or document how to disable) restart. See https://docs.micronaut.io/latest/api/io/micronaut/scheduling/io/watch/FileWatchConfiguration.html#setRestart(boolean)

mikehearn committed:
React Renderer: fix checkstyle errors.

mikehearn committed:
React Renderer: document how to avoid restarts.

mikehearn commented:
@graemerocher Done, thanks. Marked this PR as ready for review.

mikehearn commented:
CI failed due to:

```
Execution failed for task ':micronaut-views-react:findBaseline'.
> Could not find a previous version for 5.4.0
```

I guess that's expected?

graemerocher commented:
yes it is expected our builds have binary compatibility verification to ensure we don't break public API from the previous version. In this case this is a new module so there is no previous version so you need to add in the Gradle build of any new modules:

```kotlin
micronautBuild {
    binaryCompatibility.enabled.set(false)
}
```

mikehearn committed:
React Renderer: disable binary compatibility check (new module)

mikehearn committed:
React Renderer: delete package-lock.json

mikehearn commented:
Now the CI failure is a bit harder to fix. I am assuming the very latest version of GraalVM, but it seems that the CI setup provides a GraalVM already which is older (17 instead of 21). Locally my tests pass, presumably because I'm using the newer GraalVM that unbundles the truffle languages. Changing the GraalVM build used by CI for the whole project might be a separate PR or project though.

graemerocher commented:
we have a 21 CI as well in the matrix, can you exclude the tests for 17?

mikehearn committed:
React Renderer: gitignore package-lock.json

mikehearn committed:
React Renderer: exclude graalvm17 from tests in CI

mikehearn commented:
I just noticed a new problem. Some rebase has brought in a change where the blocking executor is now using virtual threads (loom) by default. Unfortunately Truffle is currently incompatible with that. It's being worked on. I will update the PR to explain this in the docs and ask users to add a new executor for js instead of using the regular blocking executor. It's not very nice, but like the restriction on using the sandbox, I hope that this will go away with new versions and the instructions can then be simplified.

mikehearn committed:
React Renderer: instruct user to disable virtual threads.

graemerocher commented:
you could always define a separate executor for JS view rendering.

mikehearn commented:
I updated the docs to tell users to just disable Loom for the blocking executor. It's only a temporary limitation and should not be a big deal.

I started writing a guide, but it seems the infrastructure has a lot of dependencies on the feature being actually published. So I will go back to that after there's been a release with this in it, as then dependency resolution etc will work. 

Do I need to update the starter system for the new feature or does that happen automatically?

graemerocher commented:
no starter will be need to be updated
